hudi-client/hudi-client-common/src/main/java/org/apache/hudi/index/HoodieIndexUtils.java

```diff
@@ -245,6 +250,59 @@ private static <R> HoodieData<HoodieRecord<R>> getExistingRecords(
         .getMergedRecords().iterator());
   }
 
+  /**
+   * getExistingRecords will create records with expression payload so we overwrite the config.
+   * Additionally, we don't want to restore this value because the write will fail later on.
+   * We also need the keygenerator so we can figure out the partition path after expression payload
+   * evaluates the merge.
+   */
+  private static Option<Pair<BaseKeyGenerator, HoodieWriteConfig>> maybeGetKeygenAndUpdatedWriteConfig(HoodieWriteConfig config, HoodieTableConfig tableConfig) {
```

Comment:
can we change the signature a bit so that its easier to manage. 

Let maybeGetKeygenAndUpdatedWriteConfig return Pair<Option<BaseKeyGenerator>, HoodieWriteConfig>

first entry in the return will only be set for expression payload. 
while 2nd entry reflects the updatedWriteConfig. 
caller is expected to use this write config irrespective of wether its expression payload or not. 

for eg, 
```
    HoodieData<HoodieRecord<R>> existingRecords = getExistingRecords(globalLocations,
        keyGeneratorWriteConfigOpt.isPresent() ? keyGeneratorWriteConfigOpt.get().getRight() : config, hoodieTable);
```

we could simplify this line to 
```
    HoodieData<HoodieRecord<R>> existingRecords = getExistingRecords(globalLocations, keyGeneratorWriteConfigOpt.get().getRight(), hoodieTable);
```

And make mergeIncomingWithExistingRecord() take in Option<BaseKeyGenerator> as last arg which will only be set for expression payload. 

this makes is look cleaner. wdyt